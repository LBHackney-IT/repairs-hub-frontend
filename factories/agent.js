import { Factory } from 'fishery'

const agentUserFactory = Factory.define(() => ({
  name: 'An Agent',
  email: 'an.agent@hackney.gov.uk',
  roles: ['agent'],
  hasRole: true,
  hasAgentPermissions: true,
  hasContractorPermissions: false,
  hasContractManagerPermissions: false,
  hasAuthorisationManagerPermissions: false,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const agent = agentUserFactory.build()
