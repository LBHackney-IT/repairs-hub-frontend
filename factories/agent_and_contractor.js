import { Factory } from 'fishery'

export const agentAndContractorUserFactory = Factory.define(() => ({
  name: 'An Operative',
  email: 'an.operative@hackney.gov.uk',
  roles: ['agent', 'contractor'],
}))

export const agentAndContractor = agentAndContractorUserFactory.build()
