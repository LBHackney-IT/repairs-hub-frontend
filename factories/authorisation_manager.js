import { Factory } from 'fishery'

export const authorisationManagerUserFactory = Factory.define(() => ({
  name: 'An Authorisation Manager',
  email: 'a.authorisation_manager@hackney.gov.uk',
  roles: ['authorisation_manager'],
}))

export const authorisationManager = authorisationManagerUserFactory.build()
