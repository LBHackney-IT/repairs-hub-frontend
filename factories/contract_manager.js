import { Factory } from 'fishery'

const contractManagerUserFactory = Factory.define(() => ({
  name: 'A Contract Manager',
  email: 'a.contract_manager@hackney.gov.uk',
  roles: ['contract_manager'],
  hasRole: true,
  hasAgentPermissions: false,
  hasContractorPermissions: false,
  hasContractManagerPermissions: true,
  hasAuthorisationManagerPermissions: false,
  hasOperativePermissions: false,
  hasAnyPermissions: true,
}))

export const contractManager = contractManagerUserFactory.build()
