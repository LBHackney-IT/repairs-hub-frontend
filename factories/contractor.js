import { Factory } from 'fishery'

export const contractorUserFactory = Factory.define(() => ({
  name: 'A Contractor',
  email: 'a.contractor@hackney.gov.uk',
  roles: ['contractor'],
}))

export const contractor = contractorUserFactory.build()
