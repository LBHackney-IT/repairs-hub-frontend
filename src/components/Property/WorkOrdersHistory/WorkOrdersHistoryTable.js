import { useContext } from 'react'
import PropTypes from 'prop-types'
import { PAGE_SIZE_AGENTS } from 'src/utils/frontEndApiClient/workOrders'
import UserContext from '../../UserContext'
import WorkOrdersHistoryRow from './WorkOrdersHistoryRow'
import { Table, THead, TBody, TR, TH } from '../../Layout/Table'
import { canAccessWorkOrder } from '../../../utils/userPermissions'

const WorkOrdersHistoryTable = ({
  workOrders,
  tabName,
  pageNumber,
  loadMoreWorkOrders,
}) => {
  const { user } = useContext(UserContext)

  const moreWorkOrdersAvailable = () => {
    // TODO: Replace with a real count from the API
    const maxWorkOrders = pageNumber * PAGE_SIZE_AGENTS

    return workOrders.length >= maxWorkOrders
  }

  const renderLoadMoreWorkOrders = () => {
    if (moreWorkOrdersAvailable()) {
      return (
        <div className="page-navigation govuk-!-padding-bottom-5">
          <button
            className="govuk-button left-page-button"
            data-module="govuk-button"
            onClick={() => loadMoreWorkOrders(pageNumber + 1)}
          >
            Load more
          </button>
        </div>
      )
    }
  }

  return (
    <>
      <h2 className="lbh-heading-h2">{tabName}</h2>

      <Table className="govuk-!-margin-top-5 work-orders-history-table">
        <THead>
          <TR className="lbh-body">
            {user && canAccessWorkOrder(user) && <TH scope="col">Reference</TH>}
            <TH scope="col">Date raised</TH>
            <TH scope="col">Trade</TH>
            <TH scope="col">Status</TH>
            <TH scope="col">Description</TH>
          </TR>
        </THead>
        <TBody>
          {workOrders.map((workOrder, index) => (
            <WorkOrdersHistoryRow key={index} {...workOrder} />
          ))}
        </TBody>
      </Table>
      {workOrders && renderLoadMoreWorkOrders()}
    </>
  )
}

WorkOrdersHistoryTable.propTypes = {
  tabName: PropTypes.string.isRequired,
  workOrders: PropTypes.arrayOf(
    PropTypes.shape({
      reference: PropTypes.number,
      dateRaised: PropTypes.instanceOf(Date),
      tradeDescription: PropTypes.string,
      status: PropTypes.string,
      description: PropTypes.string,
    })
  ).isRequired,
  pageNumber: PropTypes.number,
  pageSizeAgents: PropTypes.number,
  loadMoreWorkOrders: PropTypes.func,
}

export default WorkOrdersHistoryTable
