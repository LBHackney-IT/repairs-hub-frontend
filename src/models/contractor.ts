export default interface Contractor {
  contractorReference: string
  contractorName: string
  activeContractCount: number
}