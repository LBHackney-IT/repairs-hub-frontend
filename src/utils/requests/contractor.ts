import Contractor from '../../models/contractor'
import { APIResponseError, ApiResponseType } from '../../types/requests/types'
import { formatRequestErrorMessage } from '../errorHandling/formatErrorMessage'
import { frontEndApiRequest } from '../frontEndApiClient/requests'

export const getContractor = async (
  contractorReference: string
): Promise<ApiResponseType<Contractor | null>> => {
  try {
    const contractor = await frontEndApiRequest({
      method: 'get',
      path: `/api/contractors/${contractorReference}`,
    })

    return {
      success: true,
      response: contractor,
      error: null,
    }
  } catch (e) {
    console.error('An error has occurred:', e.response)

    return {
      success: false,
      response: null,
      error: new APIResponseError(
        e.response?.status === 404
          ? `Failed to fetch contractor with reference ${contractorReference}`
          : formatRequestErrorMessage(e)
      ),
    }
  }
}
