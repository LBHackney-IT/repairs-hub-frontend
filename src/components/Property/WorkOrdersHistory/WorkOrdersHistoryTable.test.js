import { render } from '@testing-library/react'
import UserContext from '../../UserContext'
import WorkOrdersHistoryTable from './WorkOrdersHistoryTable'
import { agent } from 'factories/agent'
import { contractor } from 'factories/contractor'
import { contractManager } from 'factories/contract_manager'
import { authorisationManager } from 'factories/authorisation_manager'

describe('WorkOrdersHistoryTable component', () => {
  const props = {
    tabName: 'Work orders history',
    workOrders: [
      {
        reference: 10000012,
        dateRaised: new Date('2021-01-18T15:28:57.17811'),
        lastUpdated: null,
        priority: '1 [I] IMMEDIATE',
        property: '16 Pitcairn House  St Thomass Square',
        owner: '',
        tradeDescription: 'DOOR ENTRY ENGINEER - DE',
        description: 'This is an immediate repair description',
        propertyReference: '00012345',
        status: 'In progress',
      },
      {
        reference: 10000013,
        dateRaised: new Date('2021-01-23T16:28:57.17811'),
        lastUpdated: null,
        priority: '4 [U] URGENT',
        property: '16 Pitcairn House  St Thomass Square',
        owner: '',
        tradeDescription: 'DOOR ENTRY ENGINEER - DE',
        description: 'This is an urgent repair description',
        propertyReference: '00012345',
        status: 'In progress',
      },
    ],
  }

  describe('when logged in as an agent', () => {
    it('should render properly', () => {
      const { asFragment } = render(
        <UserContext.Provider value={{ user: agent }}>
          <WorkOrdersHistoryTable
            workOrders={props.workOrders}
            tabName={props.tabName}
            pageSize={1}
          />
        </UserContext.Provider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('when logged in as a contractor', () => {
    it('should render properly', () => {
      const { asFragment } = render(
        <UserContext.Provider value={{ user: contractor }}>
          <WorkOrdersHistoryTable
            workOrders={props.workOrders}
            tabName={props.tabName}
            pageSize={1}
          />
        </UserContext.Provider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('when logged in as a contract manager', () => {
    it('should render properly', () => {
      const { asFragment } = render(
        <UserContext.Provider value={{ user: contractManager }}>
          <WorkOrdersHistoryTable
            workOrders={props.workOrders}
            tabName={props.tabName}
            pageSize={1}
          />
        </UserContext.Provider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('when logged in as an authorisation manager', () => {
    it('should render properly', () => {
      const { asFragment } = render(
        <UserContext.Provider value={{ user: authorisationManager }}>
          <WorkOrdersHistoryTable
            workOrders={props.workOrders}
            tabName={props.tabName}
            pageSize={1}
          />
        </UserContext.Provider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
