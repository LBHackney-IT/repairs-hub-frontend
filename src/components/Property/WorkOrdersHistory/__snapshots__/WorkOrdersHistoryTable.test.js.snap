// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkOrdersHistoryTable component when logged in as a contract manager should render properly 1`] = `
<DocumentFragment>
  <h2
    class="lbh-heading-h2"
  >
    Work orders history
  </h2>
  <table
    class="govuk-table lbh-table govuk-!-margin-top-5 work-orders-history-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row lbh-body"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Reference
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date raised
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Trade
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Status
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Description
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000012"
      >
        <td
          class="govuk-table__cell"
        >
          <a
            class="lbh-link"
            href="/work-orders/10000012"
          >
            10000012
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          18 Jan 2021
          <div
            class="work-order-hours"
          >
            15:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an immediate repair description
          </p>
        </td>
      </tr>
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000013"
      >
        <td
          class="govuk-table__cell"
        >
          <a
            class="lbh-link"
            href="/work-orders/10000013"
          >
            10000013
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          23 Jan 2021
          <div
            class="work-order-hours"
          >
            16:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an urgent repair description
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`WorkOrdersHistoryTable component when logged in as a contractor should render properly 1`] = `
<DocumentFragment>
  <h2
    class="lbh-heading-h2"
  >
    Work orders history
  </h2>
  <table
    class="govuk-table lbh-table govuk-!-margin-top-5 work-orders-history-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row lbh-body"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date raised
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Trade
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Status
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Description
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000012"
      >
        <td
          class="govuk-table__cell"
        >
          18 Jan 2021
          <div
            class="work-order-hours"
          >
            15:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an immediate repair description
          </p>
        </td>
      </tr>
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000013"
      >
        <td
          class="govuk-table__cell"
        >
          23 Jan 2021
          <div
            class="work-order-hours"
          >
            16:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an urgent repair description
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`WorkOrdersHistoryTable component when logged in as an agent should render properly 1`] = `
<DocumentFragment>
  <h2
    class="lbh-heading-h2"
  >
    Work orders history
  </h2>
  <table
    class="govuk-table lbh-table govuk-!-margin-top-5 work-orders-history-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row lbh-body"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Reference
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date raised
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Trade
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Status
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Description
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000012"
      >
        <td
          class="govuk-table__cell"
        >
          <a
            class="lbh-link"
            href="/work-orders/10000012"
          >
            10000012
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          18 Jan 2021
          <div
            class="work-order-hours"
          >
            15:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an immediate repair description
          </p>
        </td>
      </tr>
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000013"
      >
        <td
          class="govuk-table__cell"
        >
          <a
            class="lbh-link"
            href="/work-orders/10000013"
          >
            10000013
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          23 Jan 2021
          <div
            class="work-order-hours"
          >
            16:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an urgent repair description
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`WorkOrdersHistoryTable component when logged in as an authorisation manager should render properly 1`] = `
<DocumentFragment>
  <h2
    class="lbh-heading-h2"
  >
    Work orders history
  </h2>
  <table
    class="govuk-table lbh-table govuk-!-margin-top-5 work-orders-history-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row lbh-body"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Reference
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date raised
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Trade
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Status
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Description
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000012"
      >
        <td
          class="govuk-table__cell"
        >
          <a
            class="lbh-link"
            href="/work-orders/10000012"
          >
            10000012
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          18 Jan 2021
          <div
            class="work-order-hours"
          >
            15:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an immediate repair description
          </p>
        </td>
      </tr>
      <tr
        class="govuk-table__row govuk-table__row--clickable lbh-body-s"
        data-ref="10000013"
      >
        <td
          class="govuk-table__cell"
        >
          <a
            class="lbh-link"
            href="/work-orders/10000013"
          >
            10000013
          </a>
        </td>
        <td
          class="govuk-table__cell"
        >
          23 Jan 2021
          <div
            class="work-order-hours"
          >
            16:28
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          DOOR ENTRY ENGINEER - DE
        </td>
        <td
          class="govuk-table__cell"
        >
          <span
            class="govuk-tag lbh-tag lbh-tag--status-in-progress"
          >
            In progress
          </span>
        </td>
        <td
          class="govuk-table__cell description"
        >
          <p
            class="truncate-description truncate-line-5"
          >
            This is an urgent repair description
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
